### ISSUE TITLE
Colormap Support

### ISSUE THREAD

**Issue by drgrib** [wcharczuk/go-chart#33]
After examining current Go chart options, I find this one to be best-in-class for aesthetics and detail. For my purposes, I'm only missing one key feature: colormaps.

Specifically, for plotting X/Y scatterplots, I have a Z axis of values associated with each X/Y point. I need that Z axis to have its values from Z-min to Z-max mapped to a color range for each point, as shown [here](https://chrisalbon.com/python/set_the_color_of_a_matplotlib.html).

A good set of colormaps is discussed [here](http://matplotlib.org/users/colormaps.html) and [here](https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html). As you can see there, the *viridis* colormap is consistently high performing for the human eye.

**Comment by wcharczuk** [comment]
seems awesome! two things come to mind with this, one enabling a function to be the provider of the color based on the (x,y) values, the second to bundle some prebuilt color map presets, i'll take a crack at both this weekend

**Comment by wcharczuk** [comment]
took a stab at this; see `_examples/scatter/main.go` for usage

**Comment by drgrib** [comment]
@wcharczuk Thanks for taking such quick action on it. I think we are getting somewhere here. The issues I see so far

1. I'm not sure if the color spectrum in `viridis.go` are the right map for the official _[viridis]_(https://bids.github.io/colormap/images/screenshots/option_d.png). It seems to be a combination of _magma_ / _inferno_ and _viridis_. Namely, because I don't think the pink/purple it has in `_examples/scatter/output.png` occurs in _viridis_.
2. I want to be able to color based only on my Z value, rather than having it tied to the radius of the dot. Is that possible with this? Maybe by providing a Z lookup then calling `chart.Viridis` with new values?

Thanks.

**Comment by drgrib** [comment]
@wcharczuk For a clarification on my 2nd point: I would like for my `DotColorProvider` to somehow have first class access to a Z value function using `x` and `y` as parameters just like `DotWidthProvider` without having to reference the radius (or have any radius manipulation involved at all) but also use `chart.Viridis` for color values. I actually think that may be possible with your current implementation but it may not be obvious to me since I am new to your project.

**Comment by wcharczuk** [comment]
@drgrib yep, to your second point they're independent of each other, you can set a static dot width and set the `DotColorProvider` to provide the colors, i'll tweak the example to show that.

the color map i added was lifted from a matlab implementation, highly likely there are errors in it, lemme take another pass at finding a more official list of colors

**Comment by wcharczuk** [comment]
also added the jet colormap as a function (vs. a lookup table like i did with viridis)

**Comment by drgrib** [comment]
@wcharczuk Oh no! Not Jet! Viridis was created [because jet led to so many misdiagnoses by doctors who used it](https://www.youtube.com/watch?v=xAoljeRJ3lU#t=50s). Seriously, I think it is best to let it die with old software and discontinue its use. Please listen to the brief statement from the matplotlib creator in my link. People may have seriously died from its use.

And thanks for clarifying on my second point. I can help you with a pull request on _viridis_ since I am familiar with it once you have implemented the example modification you spoke of that will help me understand how to use `DotColorProvider` without modifying the radius. I have my own data set that will help me verify it quickly.

**Comment by wcharczuk** [comment]
so the code i use to generate the color based on (xrange, yrange, x, y) is in `viridis.go`, i dumped a 256 sample array from a post online, have been looking into how to do it programmatically (vs. with a prebuilt array of values).

if you look in the scatter example, i'm already using a dot width separate from the color, so you should be good to go there.

**Comment by drgrib** [comment]
@wcharczuk Okay. I will be modifying the code base today. Hopefully I'll be able to get you a pull request today or tomorrow.

**Comment by drgrib** [comment]
@wcharczuk One large change that I will be proposing is two different interfaces involved in this process. I want to change the signature of `ColorProvider` to match the signature you used for `getJetColour` in `jet.go`:

```
type ColorProvider func(v, vmin, vmax float64) drawing.Color
```

These will be the interfaces implemented by color maps like `Viridis`.

I want to then create a new interface called `DotColorProvider` to assign to the `DotColorProvider` attribute of the same name in `Stlye`:

```
type DotColorProvider func(xrange, yrange Range, x, y float64) drawing.Color
```

Then the `DotColorProvider`s will call the `ColorProvider`s. Without this change, the current implementations for color maps will be hard-coded. For `Viridis`, it is currently hard-coded to dot size. For `Jet`, it is currently hard-coded to y-position. Using these two new interfaces will free the colors up to be associated with whatever values the user wishes.

**Comment by drgrib** [comment]
Implemented in #34
